allennlp/models/reading_comprehension/document_qa.py

```diff
@@ -0,0 +1,401 @@
+                                         dim=-1)
+
+        # purple "linear ReLU layer"
+        final_merged_passage = F.relu(self._merge_atten(final_merged_passage))
+
+        # Bi-GRU in the paper
+        residual_layer = self._dropout(self._residual_encoder(self._dropout(final_merged_passage), passage_mask))
+
+        self_atten_matrix = self._self_atten(residual_layer, residual_layer)
+
+        # Expand mask for self-attention
+        mask = (passage_mask.resize(extended_batch_size, passage_length, 1) *
+                passage_mask.resize(extended_batch_size, 1, passage_length))
+
+        # Mask should have zeros on the diagonal.
+        # torch.eye does not have a gpu implementation, so we are forced to use
+        # the cpu one and .cuda(). Not sure if this matters for performance.
+        eye = torch.eye(passage_length, passage_length)
+        if mask.is_cuda:
+            eye = eye.cuda()
+        self_mask = Variable(eye).resize(1, passage_length, passage_length)
+        mask = mask * (1 - self_mask)
+
+        self_atten_probs = util.last_dim_softmax(self_atten_matrix, mask)
+
+        # Batch matrix multiplication:
+        # (batch, passage_len, passage_len) * (batch, passage_len, dim) -> (batch, passage_len, dim)
+        self_atten_vecs = torch.matmul(self_atten_probs, residual_layer)
+
+        # (extended_batch_size, passage_length, embedding_dim * 3)
+        concatenated = torch.cat([self_atten_vecs, residual_layer, residual_layer * self_atten_vecs],
+                                 dim=-1)
+
+        # _merge_self_atten => (extended_batch_size, passage_length, embedding_dim)
+        residual_layer = F.relu(self._merge_self_atten(concatenated))
+
+        # print("residual", residual_layer.size())
+
+        final_merged_passage += residual_layer
+        final_merged_passage = self._dropout(final_merged_passage)
+
+        # Bi-GRU in paper
+        start_rep = self._span_start_encoder(final_merged_passage, passage_lstm_mask)
+        span_start_logits = self._span_start_predictor(start_rep).squeeze(-1)
+        span_start_probs = util.masked_softmax(span_start_logits, passage_mask)
+
+        end_rep = self._span_end_encoder(torch.cat([final_merged_passage, start_rep], dim=-1), passage_lstm_mask)
+        span_end_logits = self._span_end_predictor(end_rep).squeeze(-1)
+        span_end_probs = util.masked_softmax(span_end_logits, passage_mask)
+
+        span_start_logits = util.replace_masked_values(span_start_logits, passage_mask, -1e7)
+        span_end_logits = util.replace_masked_values(span_end_logits, passage_mask, -1e7)
+
+        paragraph_span_start_logits = unsquash(span_start_logits, batch_size, num_paragraphs)
+        paragraph_span_end_logits = unsquash(span_end_logits, batch_size, num_paragraphs)
+
+        best_paragraph_word_span = self._get_best_span(paragraph_span_start_logits, paragraph_span_end_logits)
```

Comment:
Do you compare using logits instead of probabilities because they will compare across paragraph boundaries better than the the probabilities will?

Reply:
For getting the argmax here, it doesn't matter whether you use logits or probabilities, except that the logic assumes logits (it uses `val1 + val2` instead of `val1 * val2`).  When comparing across paragraphs, using logits instead of probabilities is important, for the reasons discussed in the paper.